### ISSUE TITLE
Defining package replacements within an app

### ISSUE THREAD

**Issue by subesokun** [jspm/jspm-cli#646]
Hi,

I'm wondering if it would be a good idea to be able to define "package replacements" within an application i.e. to be able to tell jspm that a given package shall be replaced by another compatible package during installation. Most likely the current recommended solution would be linking but let me explain a bit my current situation:

* Many of our packages are depending on a `pkg-core` package and this package gets also used in production environment so adding there changes is critical and must be well planned
* On the other hand we've some pre-development projects where we want to explore new features which also could introduce some breaking changes to `pkg-core` or any other existing package
* Increasing the major version of the `pkg-core` and add there our new features is not really an option as we're not sure if these changes should really land in the master
 * Also in case there is a planned breaking change in `pkg-core`'s master we can't introduce this breaking change as we need to jump then to next major version where we've this experimental features which shouldn't land yet into master in case we follow the strategy mentioned above (hopefully you can follow me)
* As this `pkg-core` gets used by other packages it can appear at any level of the dependency tree

So why linking is here probably not the best solution is in our case because:

* We've developers working on different features in this pre-development project and some of them don't really know about this changes in `pkg-core`. So it would be complicated to let every developer know which package he's to clone manually and link into his application in case this modified package is mandatory in order to get the application running
* Also it could happen that they forget to pull the latest changes from the cloned repos as they're not actively working on them, so there might be a great variety of used `pkg-core` versions and bugs are hard to pin down. Ultimately they just want to do a simple `jspm install` and build+test the app so that they can implement their features
* Further more we've a CI system and it would be very tricky to tell this CI from where it has to clone the extra experimental repos, how to build them locally and into which version it has to link them
* Also I can't use the new `override` feature as this will have only influence on the direct dependencies or?

In summary this is what made me think that it would be helpful to be able to define in the app's `package.json` kind of replacement rule set which declares which version range of a given package shall be replaced (in the complete dependency tree) by another package e.g.:
```json
"replace" : {
  "endp:pkg-core@^1.0.0": "end:pkg-core-exp1@^1.0.0"
}
```

Do you think that such feature would make sense or is there another approach you'd recommend to solve this problematic?

**Comment by subesokun** [comment]
ugh sorry again, maybe the current implemented override can solve this issue. I probably didn't know yet enough about them. #398
Would be great if you could give me a heads up that the current override can tackle this issue. I'll have also once more a look on them.

**Comment by brianfeister** [comment]
Also looking for documentation on the current state of `override` in `package.json` - did you figure this out @subesokun ?

**Comment by subesokun** [comment]
@brianfeister After giving some more thoughts on it I think package overrides can't solve this issue. In fact you'd need to define a package overwrite definition for every single package that consumes the package you'd like to replace. Depending on how many packages are affected this might end in a quite big package overwrite definition which is super hard to maintain.

The idea of a package replacement is more like a delta update of the dependency map. I.e. replace in every dependency map package A against B but keep the rest as it is.

**Comment by brianfeister** [comment]
Ah, makes sense, my need is a bit different, and it turns out that just doing `jspm install {package} -o "{ format: 'global' }"` gets me what I need. Thanks for the follow-up!

**Comment by subesokun** [comment]
Sure no problem.

Another option would be of course manually editing the config.js and replace there the packages. This would allow others to perform a simple jspm install in case the config.js is under version control. But then again you'd loose the ability to perform a jspm update as it reverts all manual changes... and if you update often this might the troublesome.

**Comment by subesokun** [comment]
Actually this "package replacement" is now possible via the new `map` configuration
```json
{ "map": { "pkg-core": "pkg-core-mod" } }
```
The only question left is if you can modify also the version range like `{ "map": { "pkg-core": "pkg-core-mod@^2.0.0" } }`? Or if it's possible to do even something fancy like `{ "map": { "pkg-core@^1.0.0": "pkg-core-mod@^2.0.0" } }` to make the mapping more restrictive.
